### PR TITLE
Added a cache from (orgId, userId) => permissions

### PR DESCRIPTION
**Pull request by ryanpbrewster**
Kind of tricky to invalidate the cache for only one particular org. When an
org's settings change, we need to invalidate `(orgId, *)`. I implemented a
two-cache strategy.
  The first cache is `Id[Organization] => Int`
  The second cache is `(Id[Organization], Int, Id[User]) => Set[OrganizationPermission]`

To go from `(Id[Organization], Id[User]) => Set[OrganizationPermission]`, first
hit cache #1 to get the `Int` for that orgId, then use that "namespace key" to hit
cache #2.

To invalidate an entire org, just drop the orgId key from cache #1

### CODE REVIEW

**Comment by ryanpbrewster** [comment]
@camhashemi @andrewconner @eishay

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala [hunk 1]
```diff
@@ -1,14 +1,19 @@
 package com.keepit.commanders
 
 import com.google.inject.{ ImplementedBy, Inject, Singleton }
+import com.keepit.common.cache.{ JsonCacheImpl, FortyTwoCachePlugin, CacheStatistics, Key }
 import com.keepit.common.db.Id
 import com.keepit.common.db.slick.DBSession.RSession
 import com.keepit.common.db.slick.Database
 import com.keepit.common.healthcheck.AirbrakeNotifier
-import com.keepit.common.logging.Logging
+import com.keepit.common.logging.{ AccessLog, Logging }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala:9]: don't know why this is here, maybe unneeded?

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala [hunk 2]
```diff
@@ -33,9 +40,10 @@ class PermissionCommanderImpl @Inject() (
     implicit val executionContext: ExecutionContext) extends PermissionCommander with Logging {
 
   def getOrganizationPermissions(orgId: Id[Organization], userIdOpt: Option[Id[User]])(implicit session: RSession): Set[OrganizationPermission] = {
-    // TODO(ryan): this needs to look in orgPermissionsCache by (orgId, userIdOpt) to see if the permissions are cached
-    // If not, it should compute them directly and add them to the cache
-    computeOrganizationPermissions(orgId, userIdOpt)
+    val orgPermissionsNamespace = orgPermissionsNamespaceCache.getOrElse(OrganizationPermissionsNamespaceKey(orgId))(OrganizationPermissionsNamespace(orgId, Random.nextInt()))
+    orgPermissionsCache.getOrElse(OrganizationPermissionsKey(orgPermissionsNamespace, userIdOpt)) {
+      computeOrganizationPermissions(orgId, userIdOpt)
+    }
```
Comment (ryanpbrewster) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala:57]: I considered adding a comment here for clarity. Thoughts?

**Comment by andrewconner** [comment]
@leogrim

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala (hunk 1 above)
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala:9]: used by the cache

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala [hunk 3]
```diff
@@ -145,5 +153,28 @@ class PermissionCommanderImpl @Inject() (
       OrganizationPermission.MANAGE_PLAN
     )
   }
+}
+
+case class OrganizationPermissionsNamespace(orgId: Id[Organization], value: Int)
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala]: adding the PR description here would be helpful to somebody seeing this for the first time

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala [hunk 4]
```diff
@@ -145,5 +153,28 @@ class PermissionCommanderImpl @Inject() (
...
+    (__ \ 'value).format[Int]
+  )(OrganizationPermissionsNamespace.apply, unlift(OrganizationPermissionsNamespace.unapply))
+}
+case class OrganizationPermissionsNamespaceKey(orgId: Id[Organization]) extends Key[OrganizationPermissionsNamespace] {
+  override val version = 1
+  val namespace = "org_permissions_namespace"
+  def toKey(): String = orgId.id.toString
+}
+class OrganizationPermissionsNamespaceCache(stats: CacheStatistics, accessLog: AccessLog, innermostPluginSettings: (FortyTwoCachePlugin, Duration), innerToOuterPluginSettings: (FortyTwoCachePlugin, Duration)*)
+  extends JsonCacheImpl[OrganizationPermissionsNamespaceKey, OrganizationPermissionsNamespace](stats, accessLog, innermostPluginSettings, innerToOuterPluginSettings: _*)
```
Comment (camhashemi) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/PermissionCommander.scala]: could we just store the `value` as the value and prepend the `orgId` outside of the cache hit? that way we can use a `PrimitiveCacheImpl[OrganizationPermissionsNameSpaceKey, Int]`, which should have better performance.

**Comment by camhashemi** [comment]
looks great! we went from having denormalized permissions in three db columns to having one (well, two) caches for everything.

**Pull request merged**